File: packages/stera_recorder/android/src/main/kotlin/open/fpvlabs/stera/ar_recorder/coordination/FrameProcessingCoordinator.kt
package open.fpvlabs.stera.ar_recorder.coordination

import open.fpvlabs.stera.ar_recorder.encoding.FrameSampler

class FrameProcessingCoordinator(
    private val frameSampler: FrameSampler
) {
    data class EncoderIntervalStats(
        val samples: Long,
        val meanNs: Double,
        val m2Ns: Double,
        val lastSubmittedTimestampNs: Long
    )

    data class ArcoreCadenceStats(
        val samples: Long,
        val meanIntervalNs: Double,
        val m2IntervalNs: Double,
        val lastArcoreTimestampNsForEstimate: Long
    )

    data class SamplerSyncResult(
        val totalExpectedFrameCount: Long,
        val deterministicFrameIndex: Long,
        val sessionStartTimestampNs: Long
    )

    fun syncSamplerStats(): SamplerSyncResult {
        val stats = frameSampler.getDriftStats()
        return SamplerSyncResult(
            totalExpectedFrameCount = stats["totalExpectedFrameCount"] as? Long ?: 0L,
            deterministicFrameIndex = stats["deterministicFrameIndex"] as? Long ?: 0L,
            sessionStartTimestampNs = stats["sessionStartTimestampNs"] as? Long ?: Long.MIN_VALUE
        )
    }

    fun updateEncoderIntervalStats(
        timestampNs: Long,
        lastSubmittedTimestampNs: Long,
        encoderIntervalSamples: Long,
        encoderIntervalMeanNs: Double,
        encoderIntervalM2Ns: Double
    ): EncoderIntervalStats {
        if (lastSubmittedTimestampNs == Long.MIN_VALUE) {
            return EncoderIntervalStats(
                samples = encoderIntervalSamples,
                meanNs = encoderIntervalMeanNs,
                m2Ns = encoderIntervalM2Ns,
                lastSubmittedTimestampNs = timestampNs
            )
        }
        val intervalNs = (timestampNs - lastSubmittedTimestampNs).toDouble()
        val samples = encoderIntervalSamples + 1
        val delta = intervalNs - encoderIntervalMeanNs
        val mean = encoderIntervalMeanNs + delta / samples.toDouble()
        val delta2 = intervalNs - mean
        val m2 = encoderIntervalM2Ns + (delta * delta2)
        return EncoderIntervalStats(samples, mean, m2, timestampNs)
    }

    fun updateArcoreCadenceEstimate(
        arcoreTimestampNs: Long,
        lastArcoreTimestampNsForEstimate: Long,
        arcoreCadenceSamples: Long,
        arcoreCadenceMeanIntervalNs: Double,
        arcoreCadenceM2IntervalNs: Double
    ): ArcoreCadenceStats {
        if (lastArcoreTimestampNsForEstimate != Long.MIN_VALUE && arcoreTimestampNs > lastArcoreTimestampNsForEstimate) {
            val intervalNs = (arcoreTimestampNs - lastArcoreTimestampNsForEstimate).toDouble()
            val samples = arcoreCadenceSamples + 1
            val delta = intervalNs - arcoreCadenceMeanIntervalNs
            val mean = arcoreCadenceMeanIntervalNs + delta / samples.toDouble()
            val delta2 = intervalNs - mean
            val m2 = arcoreCadenceM2IntervalNs + (delta * delta2)
            return ArcoreCadenceStats(samples, mean, m2, arcoreTimestampNs)
        }
        return ArcoreCadenceStats(
            samples = arcoreCadenceSamples,
            meanIntervalNs = arcoreCadenceMeanIntervalNs,
            m2IntervalNs = arcoreCadenceM2IntervalNs,
            lastArcoreTimestampNsForEstimate = arcoreTimestampNs
        )
    }

}

File: packages/stera_recorder/ios/stera_recorder/Sources/stera_recorder/coordination/FrameProcessingCoordinator.swift
import Foundation

class FrameProcessingCoordinator {

    private let frameSampler: FrameSampler

    init(frameSampler: FrameSampler) {
        self.frameSampler = frameSampler
    }

    struct EncoderIntervalStats {
        let samples: Int64
        let meanNs: Double
        let m2Ns: Double
        let lastSubmittedTimestampNs: Int64
    }

    struct ArkitCadenceStats {
        let samples: Int64
        let meanIntervalNs: Double
        let m2IntervalNs: Double
        let lastArkitTimestampNsForEstimate: Int64
    }

    struct SamplerSyncResult {
        let totalExpectedFrameCount: Int64
        let deterministicFrameIndex: Int64
        let sessionStartTimestampNs: Int64
    }

    func syncSamplerStats() -> SamplerSyncResult {
        let stats = frameSampler.getDriftStats()
        return SamplerSyncResult(
            totalExpectedFrameCount: stats["totalExpectedFrameCount"] as? Int64 ?? 0,
            deterministicFrameIndex: stats["deterministicFrameIndex"] as? Int64 ?? 0,
            sessionStartTimestampNs: stats["sessionStartTimestampNs"] as? Int64 ?? Int64.min
        )
    }

    func updateEncoderIntervalStats(
        timestampNs: Int64,
        lastSubmittedTimestampNs: Int64,
        encoderIntervalSamples: Int64,
        encoderIntervalMeanNs: Double,
        encoderIntervalM2Ns: Double
    ) -> EncoderIntervalStats {
        if lastSubmittedTimestampNs == Int64.min {
            return EncoderIntervalStats(
                samples: encoderIntervalSamples,
                meanNs: encoderIntervalMeanNs,
                m2Ns: encoderIntervalM2Ns,
                lastSubmittedTimestampNs: timestampNs
            )
        }
        let intervalNs = Double(timestampNs - lastSubmittedTimestampNs)
        let samples = encoderIntervalSamples + 1
        let delta = intervalNs - encoderIntervalMeanNs
        let mean = encoderIntervalMeanNs + delta / Double(samples)
        let delta2 = intervalNs - mean
        let m2 = encoderIntervalM2Ns + (delta * delta2)
        return EncoderIntervalStats(samples: samples, meanNs: mean, m2Ns: m2, lastSubmittedTimestampNs: timestampNs)
    }

    func updateArkitCadenceEstimate(
        arkitTimestampNs: Int64,
        lastArkitTimestampNsForEstimate: Int64,
        arkitCadenceSamples: Int64,
        arkitCadenceMeanIntervalNs: Double,
        arkitCadenceM2IntervalNs: Double
    ) -> ArkitCadenceStats {
        if lastArkitTimestampNsForEstimate != Int64.min && arkitTimestampNs > lastArkitTimestampNsForEstimate {
            let intervalNs = Double(arkitTimestampNs - lastArkitTimestampNsForEstimate)
            let samples = arkitCadenceSamples + 1
            let delta = intervalNs - arkitCadenceMeanIntervalNs
            let mean = arkitCadenceMeanIntervalNs + delta / Double(samples)
            let delta2 = intervalNs - mean
            let m2 = arkitCadenceM2IntervalNs + (delta * delta2)
            return ArkitCadenceStats(samples: samples, meanIntervalNs: mean, m2IntervalNs: m2, lastArkitTimestampNsForEstimate: arkitTimestampNs)
        }
        return ArkitCadenceStats(
            samples: arkitCadenceSamples,
            meanIntervalNs: arkitCadenceMeanIntervalNs,
            m2IntervalNs: arkitCadenceM2IntervalNs,
            lastArkitTimestampNsForEstimate: arkitTimestampNs
        )
    }
}
